File: InterviewPreparationKit/Pairs.js
// https://www.hackerrank.com/challenges/pairs/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function binarysearch(arr, left, right, val) {
    if (left > right) {
        return -1;
    }

    let mid = parseInt((left + right) / 2);

    if (arr[mid] == val) {
        return mid;
    }
    if (arr[mid] > val) {
        return binarysearch(arr, left, mid - 1, val);
    }
    if (arr[mid] < val) {
        return binarysearch(arr, mid + 1, right, val);
    }

    return -1;
}
    
// Complete the pairs function below.
function pairs(k, arr) {
    arr.sort(function(a, b) {
        return a - b;
    });
    let result = 0;

    for (let i = 0; i < arr.length - 1; i++) {
        if (binarysearch(arr, 0, arr.length - 1, arr[i] + k) >= 0) {
            result++;
        }
    }

    return result;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const nk = readLine().split(' ');

    const n = parseInt(nk[0], 10);

    const k = parseInt(nk[1], 10);

    const arr = readLine().split(' ').map(arrTemp => parseInt(arrTemp, 10));

    let result = pairs(k, arr);

    ws.write(result + "\n");

    ws.end();
}

File: InterviewPreparationKit/Pairs.c
// https://www.hackerrank.com/challenges/pairs/problem
#include <assert.h>
#include <limits.h>
#include <math.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* readline();
char** split_string(char*);

int compare(const void *a, const void *b) {
    int arg1 = *(const int*) a;
    int arg2 = *(const int*) b;
 
    if (arg1 < arg2) {
        return -1;
    }
    else if (arg1 > arg2) {
        return 1;
    }
    else {
        return 0;
    }
}

int binarysearch(int* arr, int left, int right, int val) {
    if (left > right) {
        return -1;
    }

    int mid = (left + right) / 2;

    if (arr[mid] == val) {
        return mid;
    }
    else if (arr[mid] > val) {
        return binarysearch(arr, left, mid - 1, val);
    }
    else {
        return binarysearch(arr, mid + 1, right, val);
    }
}

// Complete the pairs function below.
int pairs(int k, int arr_count, int* arr) {
    qsort(arr, arr_count, sizeof(int), compare);
    int result = 0;

    for (int i = 0; i < arr_count - 1; i++) {
        if (binarysearch(arr, 0, arr_count - 1, arr[i] + k) >= 0) {
            result++;
        }
    }

    return result;
}

int main()
{
    FILE* fptr = fopen(getenv("OUTPUT_PATH"), "w");

    char** nk = split_string(readline());

    char* n_endptr;
    char* n_str = nk[0];
    int n = strtol(n_str, &n_endptr, 10);

    if (n_endptr == n_str || *n_endptr != '\0') { exit(EXIT_FAILURE); }

    char* k_endptr;
    char* k_str = nk[1];
    int k = strtol(k_str, &k_endptr, 10);

    if (k_endptr == k_str || *k_endptr != '\0') { exit(EXIT_FAILURE); }

    char** arr_temp = split_string(readline());

    int* arr = malloc(n * sizeof(int));

    for (int i = 0; i < n; i++) {
        char* arr_item_endptr;
        char* arr_item_str = *(arr_temp + i);
        int arr_item = strtol(arr_item_str, &arr_item_endptr, 10);

        if (arr_item_endptr == arr_item_str || *arr_item_endptr != '\0') { exit(EXIT_FAILURE); }

        *(arr + i) = arr_item;
    }

    int arr_count = n;

    int result = pairs(k, arr_count, arr);

    fprintf(fptr, "%d\n", result);

    fclose(fptr);

    return 0;
}

char* readline() {
    size_t alloc_length = 1024;
    size_t data_length = 0;
    char* data = malloc(alloc_length);

    while (true) {
        char* cursor = data + data_length;
        char* line = fgets(cursor, alloc_length - data_length, stdin);

        if (!line) { break; }

        data_length += strlen(cursor);

        if (data_length < alloc_length - 1 || data[data_length - 1] == '\n') { break; }

        size_t new_length = alloc_length << 1;
        data = realloc(data, new_length);

        if (!data) { break; }

        alloc_length = new_length;
    }

    if (data[data_length - 1] == '\n') {
        data[data_length - 1] = '\0';
    }

    data = realloc(data, data_length);

    return data;
}

char** split_string(char* str) {
    char** splits = NULL;
    char* token = strtok(str, " ");

    int spaces = 0;

    while (token) {
        splits = realloc(splits, sizeof(char*) * ++spaces);
        if (!splits) {
            return splits;
        }

        splits[spaces - 1] = token;

        token = strtok(NULL, " ");
    }

    return splits;
}
